docs/core/app3.md

```diff
@@ -0,0 +1,324 @@
+// gas to be used by the transaction. The ratio yields an effective "gasprice",
+// which must be above some miminum to be accepted into the mempool.
+type StdFee struct {
+	Amount sdk.Coins `json:"amount"`
+	Gas    int64     `json:"gas"`
+}
+```
+
+The fee must be paid by the first signer. This allows us to quickly check if the
+transaction fee can be paid, and reject the transaction if not.
+
+## Signing
+
+The `StdTx` supports multiple messages and multiple signers.
+To sign the transaction, each signer must collect the following information:
+
+- the ChainID (TODO haven't mentioned this yet)
+- the AccountNumber and Sequence for the given signer's account (from the
+  blockchain)
+- the transaction fee 
+- the list of transaction messages
+- an optional memo
+
+Then they can compute the transaction bytes to sign using the
+`auth.StdSignBytes` function:
+
+```go
+bytesToSign := StdSignBytes(chainID, accNum, accSequence, fee, msgs, memo)
+```
+
+Note these bytes are unique for each signer, as they depend on the particular
+signers AccountNumber, Sequence, and optional memo. To facilitate easy
+inspection before signing, the bytes are actually just a JSON encoded form of
+all the relevant information.
+
+## AnteHandler
+
+As we saw in `App2`, we can use an `AnteHandler` to authenticate transactions
+before we handle any of their internal messages. While previously we implemented
+our own simple `AnteHandler`, the `x/auth` module provides a much more advanced
+one that uses `AccountMapper` and works with `StdTx`:
+
+```go
+TODO: feekeeper :(
+app.SetAnteHandler(auth.NewAnteHandler(accountMapper, feeKeeper))
+```
+
+The AnteHandler provided by `x/auth` enforces the following rules:
+
+- the memo must not be too big
+- the right number of signatures must be provided (one for each unique signer
+  returned by `msg.GetSigner` for each `msg`)
+- any account signing for the first-time must include a public key in the
+  StdSignature
+- the signatures must be valid when authenticated in the same order as specified
+  by the messages
+
+Note that validating
+signatures requires checking that the correct account number and sequence was
+used by each signer, as this information is required in the `StdSignBytes`.
+
+If any of the above are not satisfied, the AnteHandelr returns an error. 
+
+If all of the above verifications pass, the AnteHandler makes the following
+changes to the state:
+
+- increment account sequence by one for all signers
+- set the pubkey in the account for any first-time signers
+- deduct the fee from the first signer's account
+
+Recall that incrementing the `Sequence` prevents "replay attacks" where
+the same message could be executed over and over again. 
+
+The PubKey is required for signature verification, but it is only required in
+the StdSignature once. From that point on, it will be stored in the account.
+
+The fee is paid by the first address returned by `msg.GetSigners()` for the first `Msg`. 
```

Comment:
I see this function, but in the actual Antehandler impl, I don't see this being called? It just uses the first sig it sees from the TX.

Reply:
first sig in the tx must match first signer in the msgs